apps/aecore/src/aec_db.erl

```diff
@@ -244,13 +246,29 @@ backend_mode(<<"mnesia">> , #{persist := true } = M) -> M#{module => mnesia,
 transaction(Fun) when is_function(Fun, 0) ->
     mnesia:activity(transaction, Fun).
 
+dirty(Fun) when is_function(Fun, 0) ->
+    mnesia:activity(async_dirty, Fun).
+
 ensure_transaction(Fun) when is_function(Fun, 0) ->
     %% TODO: actually, some non-transactions also have an activity state
-    case get(mnesia_activity_state) of undefined ->
+    case get(mnesia_activity_state) of
+        undefined ->
+            transaction(Fun);
+        {_, _, non_transaction} ->
```

Comment:
I am concerned of this DB transaction handling change in this large PR. Would you please consider splitting DB `ensure_transaction` improvement as separate PR? I am proposing https://github.com/aeternity/aeternity/pull/2472

Any reasons why this point in the code is not using `mnesia:is_transaction()`?

Reply:
Sorry, I didn't read carefully and didn't see that you already started a new PR. But I'd prefer to also break the `aec_dirty/1` and `ensure_activity/2` changes, while I'm at it. See #2477 

The `mnesia:is_transaction()` function lacks sufficient resolution to work when you add dirty ops to the mix.

Reply:
#2477 merged into master